README.md
<p align="center">
 <img width="70%" src="docs/source/\_static/img/parlai.png" />
</p>

<p align="center">
   <a href="https://github.com/facebookresearch/ParlAI/blob/main/LICENSE">
    <img src="https://img.shields.io/badge/license-MIT-blue.svg" alt="CircleCI" />
  </a>
   <a href="https://pypi.org/project/parlai/">
    <img src="https://img.shields.io/pypi/v/parlai?color=blue&label=release" alt="CircleCI" />
  </a>
    <a href="https://circleci.com/gh/facebookresearch/ParlAI/tree/main">
    <img src="https://img.shields.io/circleci/build/github/facebookresearch/ParlAI/main" alt="Coverage" />
  </a>
    <a href="https://codecov.io/gh/facebookresearch/ParlAI">
    <img src="https://img.shields.io/codecov/c/github/facebookresearch/ParlAI" alt="GitHub contributors" />
  </a>
    <a href="https://img.shields.io/github/contributors/facebookresearch/ParlAI">
    <img src="https://img.shields.io/github/contributors/facebookresearch/ParlAI"/>
  </a>
    <a href="https://twitter.com/parlai_parley">
    <img src="https://img.shields.io/twitter/follow/parlai_parley?label=Twitter&style=social" alt="Twitter" />
  </a>
 </p>
 
-------------------------------------------------------------------------------------------------------------------------------------------------------

[ParlAI](http://parl.ai) (pronounced “par-lay”) is a python framework for
sharing, training and testing dialogue models, from open-domain chitchat, to
task-oriented dialogue, to visual question answering.

Its goal is to provide researchers:

- **100+ popular datasets available all in one place, with the same API**, among them [PersonaChat](https://arxiv.org/abs/1801.07243), [DailyDialog](https://arxiv.org/abs/1710.03957), [Wizard of Wikipedia](https://openreview.net/forum?id=r1l73iRqKm), [Empathetic Dialogues](https://arxiv.org/abs/1811.00207), [SQuAD](https://rajpurkar.github.io/SQuAD-explorer/), [MS MARCO](http://www.msmarco.org/), [QuAC](https://www.aclweb.org/anthology/D18-1241), [HotpotQA](https://hotpotqa.github.io/), [QACNN & QADailyMail](https://arxiv.org/abs/1506.03340), [CBT](https://arxiv.org/abs/1511.02301), [BookTest](https://arxiv.org/abs/1610.00956), [bAbI Dialogue tasks](https://arxiv.org/abs/1605.07683), [Ubuntu Dialogue](https://arxiv.org/abs/1506.08909), [OpenSubtitles](http://opus.lingfil.uu.se/OpenSubtitles.php),  [Image Chat](https://arxiv.org/abs/1811.00945), [VQA](http://visualqa.org/), [VisDial](https://arxiv.org/abs/1611.08669) and [CLEVR](http://cs.stanford.edu/people/jcjohns/clevr/). See the complete list [here](https://github.com/facebookresearch/ParlAI/blob/main/parlai/tasks/task_list.py).
- a wide set of [**reference models**](https://parl.ai/docs/agents_list.html) -- from retrieval baselines to Transformers.
- a large [zoo of **pretrained models**](https://parl.ai/docs/zoo.html) ready to use off-the-shelf
- seamless **integration of [Amazon Mechanical Turk](https://www.mturk.com/mturk/welcome)** for data collection and human evaluation
- **integration with [Facebook Messenger](https://parl.ai/docs/tutorial_chat_service.html)** to connect agents with humans in a chat interface
- a large range of **helpers to create your own agents** and train on several tasks with **multitasking**
- **multimodality**, some tasks use text and images


ParlAI is described in the following paper:
[“ParlAI: A Dialog Research Software Platform", arXiv:1705.06476](https://arxiv.org/abs/1705.06476)
or see these [more up-to-date slides](https://drive.google.com/file/d/1JfUW4AVrjSp8X8Fp0_rTTRoLxUfW0aUm/view?usp=sharing).

Follow us on [Twitter](https://twitter.com/parlai_parley) and check out our [Release
notes](https://github.com/facebookresearch/ParlAI/releases) to see the latest
information about new features & updates, and the website
[http://parl.ai](http://parl.ai) for further docs. For an archived list of updates,
check out [NEWS.md](https://github.com/facebookresearch/ParlAI/blob/main/NEWS.md).

<p align="center"><img width="90%" src="https://raw.githubusercontent.com/facebookresearch/ParlAI/main/docs/source/_static/img/parlai_example.png" /></p>

## Interactive Tutorial

For those who want to start with ParlAI now, you can try our [Colab Tutorial](https://colab.research.google.com/drive/1bRMvN0lGXaTF5fuTidgvlAl-Lb41F7AD#scrollTo=KtVz5dCUmFkN).

## Installing ParlAI

ParlAI currently requires Python3.8+ and [Pytorch](https://pytorch.org) 1.6 or higher.
Dependencies of the core modules are listed in [`requirements.txt`](https://github.com/facebookresearch/ParlAI/blob/main/requirements.txt). Some
models included (in [`parlai/agents`](https://github.com/facebookresearch/ParlAI/tree/main/parlai/agents)) have additional requirements.
We *strongly* recommend you install ParlAI in a [venv](https://docs.python.org/3/library/venv.html) or [conda](https://www.anaconda.com/) environment.

We do not support Windows at this time, but many users [report success on Windows using Python 3.8](https://github.com/facebookresearch/ParlAI/issues/3989) and issues with Python 3.9. We are happy to accept patches that improve Windows support.

**Standard Installation**

If you want to use ParlAI without modifications, you can install it with:

```bash
pip install parlai
```

**Development Installation**

Many users will want to modify some parts of ParlAI. To set up a development
environment, run the following commands to clone the repository and install
ParlAI:

```bash
git clone https://github.com/facebookresearch/ParlAI.git ~/ParlAI
cd ~/ParlAI; python setup.py develop
```

All needed data will be downloaded to `~/ParlAI/data`. If you need to clear out
the space used by these files, you can safely delete these directories and any
files needed will be downloaded again.

## Documentation

 - [Quick Start](https://parl.ai/docs/tutorial_quick.html)
 - [Basics: world, agents, teachers, action and observations](https://parl.ai/docs/tutorial_basic.html)
 - [Creating a new dataset/task](http://parl.ai/docs/tutorial_task.html)
 - [List of available tasks/datasets](https://parl.ai/docs/tasks.html)
 - [Creating a seq2seq agent](https://parl.ai/docs/tutorial_torch_generator_agent.html)
 - [List of available agents](https://parl.ai/docs/agents_list.html)
 - [Model zoo (list pretrained models)](https://parl.ai/docs/zoo.html)
 - [Running crowdsourcing tasks](http://parl.ai/docs/tutorial_crowdsourcing.html)
 - [Plug into Facebook Messenger](https://parl.ai/docs/tutorial_chat_service.html)


## Examples

A large set of scripts can be found in [`parlai/scripts`](https://github.com/facebookresearch/ParlAI/tree/main/parlai/scripts). Here are a few of them.
Note: If any of these examples fail, check the [installation section](#installing-parlai) to see if you have missed something.

Display 10 random examples from the SQuAD task
```bash
parlai display_data -t squad
```

Evaluate an IR baseline model on the validation set of the Personachat task:
```bash
parlai eval_model -m ir_baseline -t personachat -dt valid
```

Train a single layer transformer on PersonaChat (requires pytorch and torchtext).
Detail: embedding size 300, 4 attention heads,  2 epochs using batchsize 64, word vectors are initialized with fasttext and the other elements of the batch are used as negative during training.
```bash
parlai train_model -t personachat -m transformer/ranker -mf /tmp/model_tr6 --n-layers 1 --embedding-size 300 --ffn-size 600 --n-heads 4 --num-epochs 2 -veps 0.25 -bs 64 -lr 0.001 --dropout 0.1 --embedding-type fasttext_cc --candidates batch
```

## Code Organization

The code is set up into several main directories:

- [**core**](https://github.com/facebookresearch/ParlAI/tree/main/parlai/core): contains the primary code for the framework
- [**agents**](https://github.com/facebookresearch/ParlAI/tree/main/parlai/agents): contains agents which can interact with the different tasks (e.g. machine learning models)
- [**scripts**](https://github.com/facebookresearch/ParlAI/tree/main/parlai/scripts): contains a number of useful scripts, like training, evaluating, interactive chatting, ...
- [**tasks**](https://github.com/facebookresearch/ParlAI/tree/main/parlai/tasks): contains code for the different tasks available from within ParlAI
- [**mturk**](https://github.com/facebookresearch/ParlAI/tree/main/parlai/mturk): contains code for setting up Mechanical Turk, as well as sample MTurk tasks
- [**messenger**](https://github.com/facebookresearch/ParlAI/tree/main/parlai/chat_service/services/messenger): contains code for interfacing with Facebook Messenger
- [**utils**](https://github.com/facebookresearch/ParlAI/tree/main/parlai/utils): contains a wide number of frequently used utility methods
- [**crowdsourcing**](https://github.com/facebookresearch/ParlAI/tree/main/parlai/crowdsourcing): contains code for running crowdsourcing tasks, such as on Amazon Mechanical Turk
- [**chat_service**](https://github.com/facebookresearch/ParlAI/tree/main/parlai/chat_service/services/messenger): contains code for interfacing with services such as Facebook Messenger
- [**zoo**](https://github.com/facebookresearch/ParlAI/tree/main/parlai/zoo): contains code to directly download and use pretrained models from our model zoo

## Support
If you have any questions, bug reports or feature requests, please don't hesitate to post on our [Github Issues page](https://github.com/facebookresearch/ParlAI/issues).
You may also be interested in checking out our [FAQ](https://parl.ai/docs/faq.html) and
our [Tips n Tricks](https://parl.ai/docs/tutorial_tipsntricks.html).

Please remember to follow our [Code of Conduct](https://github.com/facebookresearch/ParlAI/blob/main/CODE_OF_CONDUCT.md).

## Contributing
We welcome PRs from the community!

You can find information about contributing to ParlAI in our
[Contributing](https://github.com/facebookresearch/ParlAI/blob/main/CONTRIBUTING.md)
document.


## The Team
ParlAI is currently maintained by Moya Chen, Emily Dinan, Dexter Ju, Mojtaba
Komeili, Spencer Poff, Pratik Ringshia, Stephen Roller, Kurt Shuster,
Eric Michael Smith, Megan Ung, Jack Urbanek, Jason Weston, Mary Williamson,
and Jing Xu. Kurt Shuster is the current Tech Lead.

Former major contributors and maintainers include Alexander H. Miller, Margaret
Li, Will Feng, Adam Fisch, Jiasen Lu, Antoine Bordes, Devi Parikh, Dhruv Batra,
Filipe de Avila Belbute Peres, Chao Pan, and Vedant Puri.

## Citation

Please cite the [arXiv paper](https://arxiv.org/abs/1705.06476) if you use ParlAI in your work:

```
@article{miller2017parlai,
  title={ParlAI: A Dialog Research Software Platform},
  author={{Miller}, A.~H. and {Feng}, W. and {Fisch}, A. and {Lu}, J. and {Batra}, D. and {Bordes}, A. and {Parikh}, D. and {Weston}, J.},
  journal={arXiv preprint arXiv:{1705.06476}},
  year={2017}
}
```

## License
ParlAI is MIT licensed. See the **[LICENSE](https://github.com/facebookresearch/ParlAI/blob/main/LICENSE)** file for details.


NEWS.md
## News

**Archive Information**
We are now making most of our major announcements on [Twitter](https://twitter.com/parlai_parley) and in our
[Release Notes](https://github.com/facebookresearch/ParlAI/releases).

### Older news

2020-07-21: We have added a new `parlai` super command, complete with [autocompletion](https://parl.ai/docs/tutorial_tipsntricks.html#command-line-tool).

2020-07-17: We've added a new tutorial showing how to [speed up training](https://parl.ai/docs/tutorial_fast.html).

2020-04-28: [BlenderBot](https://parl.ai/projects/recipes/) has been released, including 2.7B and 9.4B parameter models. It received some attention in the media, including [Engadget](https://www.engadget.com/facebook-chatbot-ai-open-source-blender-150001320.html), [Technology Review](https://www.technologyreview.com/2020/04/29/1000795/facebook-ai-chatbot-blender-beats-google-meena/), [Fortune](https://fortune.com/2020/04/29/facebook-creates-the-most-human-chatbot-yet/), and [Wired](https://www.wired.com/story/new-chatbot-tries-artificial-empathy/).

2020-02-04: [GPT2 fine-tuning for dialogue](https://github.com/facebookresearch/ParlAI/tree/main/parlai/agents/hugging_face)  now supported in ParlAI!

2020-01-17: [All-in-One Image-Grounded Conversational Agents project](https://twitter.com/parlai_parley/status/1218204771859017732) built in ParlAI 

2020-01-15: [Seb Ruder's 10 ML & NLP Research Highlights of 2019 post](https://ruder.io/research-highlights-2019/) mentions at least two works built with ParlAI.

2019-12-19: [Amazon's TopicChat dataset uses ParlAI](https://twitter.com/parlai_parley/status/1207716855525367808).

2019-12-11: [OneCommon Corpus is now available in ParlAI](https://twitter.com/futsaludy/status/1204661465967259648). Thanks to our colleague Takato Yamazaki

2019-11-10: [Dialogue unlikelihood project](https://parl.ai/projects/dialogue_unlikelihood/) for improving consistency in dialogue generation. 

2019-11-09: The [Gender bias project page](https://parl.ai/projects/genderation_bias/) mitigates bias in dialogue generation.

2019-11-09: The [Dodeca dialogue project page](https://parl.ai/projects/dodecadialogue/) is up with paper and leaderboard -- multi-tasking on many dialogue tasks with strong results!

2019-09-16: The [Dialogue Safety project page](https://parl.ai/projects/dialogue_safety/) is updated with pretrained models, data, and an interactive demonstration.

2019-06-12: The [Personality-Captions project page](http://parl.ai/projects/personality_captions/) is updated with pretrained models, specialized model code, and an interactive demo.

2019-06-12: Added [HotpotQA](https://github.com/facebookresearch/ParlAI/tree/main/parlai/tasks/quac) as another dataset.

2019-05-15: Added [QuAC](https://github.com/facebookresearch/ParlAI/tree/main/parlai/tasks/quac), Question Answering in Context dataset, and [COQA](https://github.com/facebookresearch/ParlAI/tree/main/parlai/tasks/coqa), Conversational Question Answering Challenge datasets.

2019-05-03: The [What makes a good Conversation project page](https://parl.ai/projects/controllable_dialogue/) is now available with pretrained models.

2019-03-15: The [Wizard of Wikipedia project page](http://parl.ai/projects/wizard_of_wikipedia/) is updated with pretrained models and specialized model code.

2019-03-09: Added [LIGHT](http://parl.ai/projects/light) text adventure game research platform for learning to speak and act. [[press]](https://venturebeat.com/2019/03/08/facebook-ai-researchers-create-a-text-based-adventure-to-study-how-ai-speak-and-act/)

2019-03-06: Added [Self-feeding Chatbot](http://parl.ai/projects/self_feeding) for leveraging user textual feedback to improve the chatbot's abilities. [[press]](https://venturebeat.com/2019/01/17/facebook-and-stanford-researchers-design-a-chatbot-that-learns-from-its-mistakes/)

2019-02-07: Added [BERT Ranker agents](https://github.com/facebookresearch/ParlAI/tree/main/parlai/agents/bert_ranker), several variations of a ranking model based on the pretrained language model BERT.

2019-01-16: ParlAI has been relicensed under the MIT open source license.

2018-12-13: Added [Daily Dialog](https://github.com/facebookresearch/ParlAI/blob/main/parlai/tasks/dailydialog/agents.py), an open-domain daily dialogue dataset.

2018-11-05: Added [Wizard of Wikipedia](http://parl.ai/projects/wizard_of_wikipedia/), a dataset for knowledge-powered conversation.

2018-11-02: Added [Image-Chat](https://klshuster.github.io/image_chat/), a dataset for engaging personality-conditioned dialogue grounded in images.

2018-10-25: Added [Personality-Captions](https://arxiv.org/abs/1810.10665), a dataset for engaging image captioning via personality.

2018-08-29: Added new cleaner version of seq2seq model with new TorchAgent parent class, along with folder (parlai/legacy_agents) for deprecated model code

2018-07-17: Added [Qangaroo](http://qangaroo.cs.ucl.ac.uk/) (a.k.a. WikiHop and MedHop), two reading comprehension datasets with multiple hops, and [SQuAD 2](https://rajpurkar.github.io/SQuAD-explorer/).

2018-05-22: Two new tasks added: [COCO Image Captioning](http://cocodataset.org/#captions-2015) and [Flickr30k Entities](http://web.engr.illinois.edu/~bplumme2/Flickr30kEntities/)

2018-04-13: [NIPS ConvAI2 competition!](http://convai.io/) Train Dialogue Agents to chat about personal interests and get to know their dialogue partner -- using the PersonaChat dataset as a training source, with data and baseline code in ParlAI. Competition starts now! Ends September 1st.

2018-03-13: Added [ParlAI-Messenger](http://parl.ai/static/docs/messenger.html), a new method for connecting human agents to a world in ParlAI using Facebook Messenger. Host your bots on Facebook Messenger to expose them to a broad audience!


microbench.py
#!/usr/bin/env python

import timeit
import torch
import apex.fp16_utils.fp16_optimizer as fp16_optimizer
import apex.optimizers.fused_adam as fused_adam
import argparse
import torch.nn.functional as F
import torch.utils.data
from parlai.core.utils import AttrDict
from parlai.agents.transformer.modules import TransformerEncoder as ParlaiTransformerEncoder
from fairseq.models.transformer import TransformerEncoder as FSTransformerEncoder

class SimpleModel(torch.nn.Module):
    def __init__(self, embedding, esz, ffn):
        super().__init__()
        self.embedding = embedding
        self.lin = torch.nn.Linear(esz, ffn)
        self.lin2 = torch.nn.Linear(ffn, esz)

    def forward(self, x):
        return self.lin2(F.relu(self.lin(self.embedding(x))))

NUM_BATCHES = 300
BS = 32
SEQLEN = 512
VOCAB = 16384
FFN = 1024
ESZ = 256
N_LAYERS = 4
N_HEADS = 2

ap = argparse.ArgumentParser()
ap.add_argument('--unsafe', action='store_true')
ap.add_argument('--bs', default=BS, type=int, help='batch size')
ap.add_argument('--type', default='parlai', choices={'parlai', 'fairseq', 'linear'})
ap.add_argument('--seqlen', default=SEQLEN, type=int, help='Sequence length')
ap.add_argument('--esz', default=ESZ, type=int, help='embedding size')
ap.add_argument('--ffn', default=FFN, type=int, help='ffn size')
ap.add_argument('--layers', type=int, default=N_LAYERS, help='mha layers')
ap.add_argument('--heads', type=int, default=N_HEADS, help='mha heads')
ap.add_argument('--vocab', type=int, default=VOCAB, help='vocab size')
ap.add_argument('--num', type=int, default=NUM_BATCHES, help='number of iterations')

args = ap.parse_args()


def build_model(args):
    embeddings = torch.nn.Embedding(args.vocab, args.esz, 0)
    if args.type == 'parlai':
        model = ParlaiTransformerEncoder(
            args.heads,  # heads
            args.layers,  # layers
            args.esz,  # embedding size
            args.ffn,  # ffn size
            args.vocab,  # vocab size
            embedding=embeddings,
            learn_positional_embeddings=False,
            n_positions=args.seqlen,
            reduction=False,
        )
    elif args.type == 'fairseq':
        model = FSTransformerEncoder(AttrDict({
            'dropout': 0.0,
            'max_source_positions': args.seqlen - 1,  # fairseq adds 1
            'encoder_learned_pos': False,
            'encoder_normalize_before': False,
            'encoder_layers': args.layers,
            'encoder_attention_heads': args.heads,
            'encoder_relu_dropout': 0.0,
            'encoder_ffn_embed_dim': args.ffn,
            'no_token_positional_embeddings': False,
            'encoder_embed_dim': args.esz,
            'attention_dropout': 0.0,
            'relu_dropout': 0.0,
        }), None, embeddings)
    elif args.type == 'linear':
        model = SimpleModel(embeddings, args.esz, args.ffn)
    return model, embeddings



class RandomSentences(torch.utils.data.Dataset):
    def __init__(self, args):
        self.seqlen = args.seqlen
        self.vocab = args.vocab

    def __len__(self):
        return BS * NUM_BATCHES

    def __getitem__(self, i):
        return torch.randint(low=1, high=self.vocab - 3, size=(self.seqlen, ), dtype=torch.int64)

    @classmethod
    def collate_fn(cls, batch):
        return torch.stack(batch).cuda()



dl = torch.utils.data.DataLoader(
    RandomSentences(args),
    args.bs,
    collate_fn=RandomSentences.collate_fn,
    num_workers=0
)

def run_loop(dl, args, use_fp16, use_apex):
    model, embeddings = build_model(args)
    lens = torch.LongTensor([args.seqlen] * args.bs).cuda()
    model = model.cuda()
    if use_fp16:
        model = model.half()
    else:
        model = model.float()

    for p in model.parameters():
        for d in p.shape:
            if d % 8 != 0:
                print("Warning: parameter isn't mod 8: ", p.shape)

    if use_apex:
        opt = fused_adam.FusedAdam(model.parameters(), lr=1e-5)
        opt = fp16_optimizer.FP16_Optimizer(opt) #, dynamic_loss_scale=True)
    else:
        opt = torch.optim.Adam(model.parameters(), lr=1e-5)

    start = timeit.default_timer()
    for x in dl:
        opt.zero_grad()

        if args.type == 'parlai':
            enc, _ = model(x)
        elif args.type == 'fairseq':
            enc = model(x, lens)['encoder_out']
        elif args.type == 'linear':
            enc = model(x)

        output = F.linear(enc.view(-1, enc.size(2)), embeddings.weight)
        output = output.view(-1, output.size(-1))
        if args.unsafe:
            logits = F.log_softmax(output, dim=-1)
        else:
            logits = F.log_softmax(output.float(), dim=-1).type_as(output)
        loss = F.nll_loss(logits, x.view(-1) + 1)

        if use_apex:
            opt.backward(loss)
            opt.clip_master_grads(0.1)
        else:
            loss.backward()
            torch.nn.utils.clip_grad_norm_(model.parameters(), 0.1)
        opt.step()

    torch.cuda.synchronize()
    end = timeit.default_timer()
    print(loss.item())

    del model
    del embeddings
    return end - start

fp16_apex = run_loop(dl, args, True, True)
print("FP16, APEX: %.2f" % fp16_apex)
fp16_raw = run_loop(dl, args, True, False)
print("FP16, RAW: %.2f" % fp16_raw)
fp32 = run_loop(dl, args, False, False)
print("FP32:", fp32)
print("Speedup: %.1f" % (fp32 / fp16_apex))
